### PR TITLE
refactor RegionManager

### PR DESCRIPTION
**Pull request by bobotu**
Introduce `RegionManager` interface which supports `GetRegionFromCtx` and `CheckLeader`.

PTAL @coocood

### CODE REVIEW

File: tikv/region.go [hunk 1]
```diff
@@ -176,10 +177,147 @@ type RegionOptions struct {
 	RegionSize int64
 }
 
-type RegionManager struct {
-	storeMeta  metapb.Store
-	mu         sync.RWMutex
-	regions    map[uint64]*regionCtx
+type RegionManager interface {
+	GetRegionFromCtx(ctx *kvrpcpb.Context) (*regionCtx, *errorpb.Error)
+	CheckLeader(ctx *kvrpcpb.Context) *errorpb.Error
```
Comment (coocood) [inline tikv/region.go]: We can do `CheckLeader` inside `GetRegionFromCtx` and remove `CheckLeader` method from `RegionManager`.

File: tikv/region.go [hunk 2]
```diff
@@ -176,10 +177,147 @@ type RegionOptions struct {
...
+}
+
+func (rm *regionManager) isEpochStale(lhs, rhs *metapb.RegionEpoch) bool {
+	return lhs.GetConfVer() != rhs.GetConfVer() || lhs.GetVersion() != rhs.GetVersion()
+}
+
+type RaftRegionManager struct {
+	regionManager
+	router   raftstore.RaftstoreRouter
+	checkers map[uint64]raftstore.LeaderChecker
```
Comment (coocood) [inline tikv/region.go]: If we put `raftstore.LeaderChecker` into `regionCtx`, then we don't need to access two maps under lock.

File: tikv/region.go [hunk 3]
```diff
@@ -176,10 +177,147 @@ type RegionOptions struct {
...
+	delete(rm.regions, ctx.RegionId)
+	delete(rm.checkers, ctx.RegionId)
+	rm.mu.Unlock()
+	region.refCount.Done()
+}
+
+func (rm *RaftRegionManager) OnSplitRegion(derived *metapb.Region, regions []*metapb.Region, peers []*raftstore.PeerEventContext) {
+	rm.mu.Lock()
+	oldRegion := rm.regions[derived.Id]
+	oldRegion.waitParent()
```
Comment (coocood) [inline tikv/region.go]: Can we `waitParent` outside of the lock?

**Comment by coocood** [comment]
The code would be cleaner if we move the split region logic out of base `regionManager`.

**Comment by bobotu** [comment]
PTAL @bobotu

**Comment by coocood** [comment]
LGTM

**Pull request merged**